src/community/modularity.c

```diff
@@ -262,28 +307,61 @@ int igraph_modularity_matrix(const igraph_t *graph,
         IGRAPH_ERROR("The resolution parameter must be non-negative", IGRAPH_EINVAL);
     }
 
-    IGRAPH_VECTOR_INIT_FINALLY(&deg, no_of_nodes);
-    if (!weights) {
-        IGRAPH_CHECK(igraph_degree(graph, &deg, igraph_vss_all(), IGRAPH_ALL,
-                                   IGRAPH_LOOPS));
-    } else {
-        IGRAPH_CHECK(igraph_strength(graph, &deg, igraph_vss_all(), IGRAPH_ALL,
-                                     IGRAPH_LOOPS, weights));
+    if (!igraph_is_directed(graph)) {
+        directed = 0;
     }
-    IGRAPH_CHECK(igraph_get_adjacency(graph, modmat, IGRAPH_GET_ADJACENCY_BOTH,
-                                      /*eids=*/ 0));
+    IGRAPH_CHECK(igraph_i_modularity_matrix_get_adjacency(graph, modmat, weights, directed));
 
-    for (i = 0; i < no_of_nodes; i++) {
-        MATRIX(*modmat, i, i) *= 2;
-    }
-    for (i = 0; i < no_of_nodes; i++) {
-        for (j = 0; j < no_of_nodes; j++) {
-            MATRIX(*modmat, i, j) -= resolution * VECTOR(deg)[i] * VECTOR(deg)[j] / 2.0 / sw;
+   if (directed) {
+        IGRAPH_VECTOR_INIT_FINALLY(&in_deg, no_of_nodes);
+        IGRAPH_VECTOR_INIT_FINALLY(&out_deg, no_of_nodes);
+        if (!weights) {
+            IGRAPH_CHECK(igraph_degree(graph, &in_deg, igraph_vss_all(), IGRAPH_IN,
+                                       IGRAPH_LOOPS));
+            IGRAPH_CHECK(igraph_degree(graph, &out_deg, igraph_vss_all(), IGRAPH_OUT,
+                                       IGRAPH_LOOPS));
+        } else {
+            IGRAPH_CHECK(igraph_strength(graph, &in_deg, igraph_vss_all(), IGRAPH_IN,
+                                         IGRAPH_LOOPS, weights));
+            IGRAPH_CHECK(igraph_strength(graph, &out_deg, igraph_vss_all(), IGRAPH_OUT,
+                                         IGRAPH_LOOPS, weights));
         }
-    }
+        /* Scaling one degree factor so every element gets scaled. */
+        scaling_factor = resolution / sw;
+        igraph_vector_scale(&out_deg, scaling_factor);
 
-    igraph_vector_destroy(&deg);
-    IGRAPH_FINALLY_CLEAN(1);
+        for (i = 0; i < no_of_nodes; i++) {
+            for (j = 0; j < no_of_nodes; j++) {
+                MATRIX(*modmat, i, j) -= VECTOR(out_deg)[i] * VECTOR(in_deg)[j];
+            }
+        }
+        igraph_vector_destroy(&in_deg);
+        igraph_vector_destroy(&out_deg);
+        IGRAPH_FINALLY_CLEAN(2);
+    } else {
+        IGRAPH_VECTOR_INIT_FINALLY(&deg, no_of_nodes);
+        if (!weights) {
+            IGRAPH_CHECK(igraph_degree(graph, &deg, igraph_vss_all(), IGRAPH_ALL,
+                                       IGRAPH_LOOPS));
+        } else {
+            IGRAPH_CHECK(igraph_strength(graph, &deg, igraph_vss_all(), IGRAPH_ALL,
+                                         IGRAPH_LOOPS, weights));
+        }
+
+        /* Scaling one degree factor so every element gets scaled. */
+        igraph_vector_copy(&deg_unscaled, &deg);
```

Comment:
If we scale with `sqrt(2*sw)` then we can avoid making a copy of the degree vector.

Reply:
I did try a square root before, but I am allowing negative weights, and then things get messy :)